File: solutions/java/code-carl/ch10_monotone_stack/T2_LC0496_NextGreaterElement.java
/**
 * Practice template for:
 * - LeetCode 496. Next Greater Element I
 *   https://leetcode.cn/problems/next-greater-element-i/
 */
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/28 17:10
 * @comment
 */
public class T2_LC0496_NextGreaterElement {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Deque<Integer> deque = new ArrayDeque<>();
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums1.length; i++) {
            map.put(nums1[i], i);
        }

        int[] res = new int[nums1.length];
        Arrays.fill(res, -1);
        for (int i = 0; i < nums2.length; i++) {
            while (!deque.isEmpty() && nums2[i] > deque.peek()) {
                Integer n = deque.pop();
                if (map.containsKey(n)) {
                    res[map.get(n)] = nums2[i];
                }
            }
            deque.push(nums2[i]);
        }

        return res;
    }
}

File: solutions/code-carl/ch10_monotone_stack/t2_lc0496_next_greater_element.py
"""LeetCode 496. Next Greater Element I
https://leetcode.cn/problems/next-greater-element-i/
"""

from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dic = {n: i for i, n in enumerate(nums1)}
        ans = [-1] * len(nums1)

        stack: List[int] = []
        for i, n in enumerate(nums2):
            while stack and n > nums2[stack[-1]]:
                n_in_nums2 = nums2[stack.pop()]
                if n_in_nums2 in dic:
                    ans[dic[n_in_nums2]] = n
            stack.append(i)

        return ans


if __name__ == "__main__":
    solution = Solution()
    assert solution.nextGreaterElement([4, 1, 2], [1, 3, 4, 2]) == [-1, 3, -1]

    assert solution.nextGreaterElement([4, 1, 2], [1, 3, 4, 2]) == [-1, 3, -1]

    print("All tests passed.")
